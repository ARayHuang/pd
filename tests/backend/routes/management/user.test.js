const request = require('supertest');
const { connect } = require('ljit-db/mongoose');
const config = require('config');
const mongoose = require('mongoose');
const {
	deleteDocuments,
	insertDocuments,
} = require('../../../../server/fixtures/index');
const { app } = require('../../../../server/management');
const LOGIN = {
	ADMIN: {
		password: '123qwe',
		username: 'admin',
	},
	MANAGER: {
		PROVIDER: {
			password: '123qwe',
			username: 'provider01',
		},
		CONSUMER: {
			password: '123qwe',
			username: 'consumer01',
		},
	},
};
const USER_ID = {
	NOT_FOUND: '5ec4ec469d87c4387d9350b0',
	INVALID: '5ec4ec469d87c4387d9350b',
	MANAGER: {
		PROVIDER: '5ec4ec469d87c4387d9350ba',
	},
	STAFF: {
		PROVIDER: '5ec4ec469d87c4387d9350bb',
	},
};
let ADMIN_COOKIE;
let PROVIDER_MANAGER_COOKIE;
let CONSUMER_MANAGER_COOKIE;

beforeAll(async () => {
	try {
		await connect(config.SERVER.MONGO.URL, { debug: false });
		await deleteDocuments();
		await insertDocuments();

		await request(app)
			.post('/api/v1/login')
			.send(LOGIN.ADMIN)
			.set('Content-Type', 'application/json')
			.set('Accept', 'application/json')
			.expect(201)
			.then(res => {
				ADMIN_COOKIE = res.headers['set-cookie'];
			});
		await request(app)
			.post('/api/v1/login')
			.send(LOGIN.MANAGER.PROVIDER)
			.set('Content-Type', 'application/json')
			.set('Accept', 'application/json')
			.expect(201)
			.then(res => {
				PROVIDER_MANAGER_COOKIE = res.headers['set-cookie'];
			});
		await request(app)
			.post('/api/v1/login')
			.send(LOGIN.MANAGER.CONSUMER)
			.set('Content-Type', 'application/json')
			.set('Accept', 'application/json')
			.expect(201)
			.then(res => {
				CONSUMER_MANAGER_COOKIE = res.headers['set-cookie'];
			});
	} catch (error) {
		throw error;
	}
});

afterAll(() => {
	return mongoose.connection.close();
});

describe('user', () => {
	describe('get /api/v1/users', () => {
		test('200 ok', () => {
			return request(app)
				.get('/api/v1/users/?type=manager')
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', ADMIN_COOKIE)
				.expect(200)
				.then(res => {
					expect(res.body).toMatchSnapshot({
						data: [
							{ id: expect.any(String) },
							{ id: expect.any(String) },
						],
					});
				});
		});
		test('422 invalid type', () => {
			return request(app)
				.get('/api/v1/users/?type=invalid')
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', ADMIN_COOKIE)
				.expect(422)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
	});

	describe('post /api/v1/users?via=admin', () => {
		test('201 created', () => {
			const body = {
				profilePictureId: '1',
				departmentType: 'provider',
				displayName: '新增主管帳號測試',
				password: '123qwe',
				username: 'createManagerTest',
				hasPermissionToAddStaff: true,
			};

			return request(app)
				.post('/api/v1/users?via=admin')
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', ADMIN_COOKIE)
				.expect(201)
				.then(res => {
					expect(res.body).toMatchSnapshot({
						id: expect.any(String),
					});
				});
		});
		test('403 forbidden', () => {
			const body = {
				profilePictureId: '1',
				departmentType: 'provider',
				displayName: '新增主管帳號測試',
				password: '123qwe',
				username: 'createManagerTest',
				hasPermissionToAddStaff: true,
			};

			return request(app)
				.post('/api/v1/users?via=admin')
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(403)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('409 created', () => {
			const body = {
				profilePictureId: '1',
				departmentType: 'provider',
				displayName: '新增主管帳號測試',
				password: '123qwe',
				username: 'admin',
				hasPermissionToAddStaff: true,
			};

			return request(app)
				.post('/api/v1/users?via=admin')
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', ADMIN_COOKIE)
				.expect(409)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('422 invalid profilePictureId', () => {
			const body = {
				profilePictureId: '0',
				departmentType: 'provider',
				displayName: '新增主管帳號測試',
				password: '123qwe',
				username: 'createManagerTest',
				hasPermissionToAddStaff: true,
			};

			return request(app)
				.post('/api/v1/users?via=admin')
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', ADMIN_COOKIE)
				.expect(422)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
	});

	describe('post /api/v1/users?via=manager', () => {
		test('201 created', () => {
			const body = {
				profilePictureId: '1',
				shiftType: 'morning',
				channelIds: ['5ec5f60b25fbb049dbd231e2'],
				displayName: '新增部門人員帳號測試',
				password: '123qwe',
				username: 'createStaffTest',
			};

			return request(app)
				.post('/api/v1/users?via=manager')
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(201)
				.then(res => {
					expect(res.body).toMatchSnapshot({
						id: expect.any(String),
					});
				});
		});
		test('403 forbidden', () => {
			const body = {
				profilePictureId: '1',
				shiftType: 'morning',
				channelIds: ['5ec5f60b25fbb049dbd231e2'],
				displayName: '新增部門人員帳號測試',
				password: '123qwe',
				username: 'createStaffTest',
			};

			return request(app)
				.post('/api/v1/users?via=manager')
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', ADMIN_COOKIE)
				.expect(403)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('409 created', () => {
			const body = {
				profilePictureId: '1',
				shiftType: 'morning',
				channelIds: ['5ec5f60b25fbb049dbd231e2'],
				displayName: '新增部門人員帳號測試',
				password: '123qwe',
				username: 'createStaffTest',
			};

			return request(app)
				.post('/api/v1/users?via=manager')
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(409)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('422 invalid profilePictureId', () => {
			const body = {
				profilePictureId: '0',
				shiftType: 'morning',
				channelIds: ['5ec5f60b25fbb049dbd231e2'],
				displayName: '新增部門人員帳號測試',
				password: '123qwe',
				username: 'createStaffTest',
			};

			return request(app)
				.post('/api/v1/users?via=manager')
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(422)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
	});

	describe('patch /api/v1/users/id=:userId?via=admin', () => {
		test('204 no content', () => {
			const body = {
				profilePictureId: '1',
			};

			return request(app)
				.patch(`/api/v1/users/id=${USER_ID.MANAGER.PROVIDER}?via=admin`)
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', ADMIN_COOKIE)
				.expect(204)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('403 forbidden', () => {
			const body = {
				profilePictureId: '1',
			};

			return request(app)
				.patch(`/api/v1/users/id=${USER_ID.MANAGER.PROVIDER}?via=admin`)
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(403)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('404 manager not found', () => {
			const body = {
				profilePictureId: '1',
			};

			return request(app)
				.patch(`/api/v1/users/id=${USER_ID.NOT_FOUND}?via=admin`)
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', ADMIN_COOKIE)
				.expect(404)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('422 invalid profilePictureId', () => {
			const body = {
				profilePictureId: '',
			};

			return request(app)
				.patch(`/api/v1/users/id=${USER_ID.MANAGER.PROVIDER}?via=admin`)
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', ADMIN_COOKIE)
				.expect(422)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
	});

	describe('patch /api/v1/users/id=:userId?via=manager', () => {
		test('204 no content', () => {
			const body = {
				profilePictureId: '1',
			};

			return request(app)
				.patch(`/api/v1/users/id=${USER_ID.STAFF.PROVIDER}?via=manager`)
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(204)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('403 forbidden', () => {
			const body = {
				profilePictureId: '1',
			};

			return request(app)
				.patch(`/api/v1/users/id=${USER_ID.STAFF.PROVIDER}?via=manager`)
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', CONSUMER_MANAGER_COOKIE)
				.expect(403)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('404 user not found', () => {
			const body = {
				profilePictureId: '1',
			};

			return request(app)
				.patch(`/api/v1/users/id=${USER_ID.NOT_FOUND}?via=manager`)
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(404)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('422 invalid profilePictureId', () => {
			const body = {
				profilePictureId: '',
			};

			return request(app)
				.patch(`/api/v1/users/id=${USER_ID.STAFF.PROVIDER}?via=manager`)
				.send(body)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(422)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
	});

	describe('get /api/v1/users/id=me', () => {
		test('200 ok', () => {
			return request(app)
				.get('/api/v1/users/id=me')
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', PROVIDER_MANAGER_COOKIE)
				.expect(200)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('401 unauthorized', () => {
			return request(app)
				.get('/api/v1/users/id=me')
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.expect(401)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
	});

	describe('get /api/v1/users/id=:userId', () => {
		test('200 ok', () => {
			return request(app)
				.get(`/api/v1/users/id=${USER_ID.MANAGER.PROVIDER}`)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', ADMIN_COOKIE)
				.expect(200)
				.then(res => {
					expect(res.body).toMatchSnapshot({
						id: expect.any(String),
					});
				});
		});
		test('404 user not found', () => {
			return request(app)
				.get(`/api/v1/users/id=${USER_ID.NOT_FOUND}`)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', ADMIN_COOKIE)
				.expect(404)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('422 invalid userId', () => {
			return request(app)
				.get(`/api/v1/users/id=${USER_ID.INVALID}`)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', ADMIN_COOKIE)
				.expect(422)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
	});

	describe('delete /api/v1/users/id=:userId', () => {
		test('204 no content', () => {
			return request(app)
				.delete(`/api/v1/users/id=${USER_ID.MANAGER.PROVIDER}`)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', ADMIN_COOKIE)
				.expect(204)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('404 user not found', () => {
			return request(app)
				.delete(`/api/v1/users/id=${USER_ID.NOT_FOUND}`)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', ADMIN_COOKIE)
				.expect(404)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
		test('422 invalid userId', () => {
			return request(app)
				.delete(`/api/v1/users/id=${USER_ID.INVALID}`)
				.set('Content-Type', 'application/json')
				.set('Accept', 'application/json')
				.set('Cookie', ADMIN_COOKIE)
				.expect(422)
				.then(res => {
					expect(res.body).toMatchSnapshot();
				});
		});
	});
});
