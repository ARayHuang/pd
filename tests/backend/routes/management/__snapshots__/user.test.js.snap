// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user delete /api/v1/users/id=:userId 204 no content 1`] = `Object {}`;

exports[`user delete /api/v1/users/id=:userId 404 user not found 1`] = `
Object {
  "code": "888.012.404",
  "message": "用户不存在",
  "type": "NotFoundError",
}
`;

exports[`user delete /api/v1/users/id=:userId 422 invalid userId 1`] = `
Object {
  "code": "888.012.422",
  "fields": Array [
    "userId",
  ],
  "message": "输入格式错误",
  "type": "RequestValidationError",
}
`;

exports[`user get /api/v1/users 200 ok 1`] = `
Object {
  "data": Array [
    Object {
      "channels": Array [
        Object {
          "id": "5ec5f60b25fbb049dbd231e2",
          "name": "频道一",
        },
      ],
      "departmentType": "consumer",
      "displayName": "接单组管理员",
      "hasPermissionToAddStaff": true,
      "id": Any<String>,
      "isOnline": false,
      "profilePictureId": "1",
      "type": "manager",
      "username": "consumer01",
    },
    Object {
      "channels": Array [
        Object {
          "id": "5ec5f60b25fbb049dbd231e2",
          "name": "频道一",
        },
      ],
      "departmentType": "provider",
      "displayName": "开单组管理员",
      "hasPermissionToAddStaff": true,
      "id": Any<String>,
      "isOnline": false,
      "profilePictureId": "1",
      "type": "manager",
      "username": "provider01",
    },
  ],
  "numOfItems": 2,
  "numOfPages": 1,
}
`;

exports[`user get /api/v1/users 422 invalid type 1`] = `
Object {
  "code": "888.012.422",
  "fields": Array [
    "type",
    "departmentType",
  ],
  "message": "输入格式错误",
  "type": "RequestValidationError",
}
`;

exports[`user get /api/v1/users/id=:userId 200 ok 1`] = `
Object {
  "channels": Array [
    Object {
      "id": "5ec5f60b25fbb049dbd231e2",
      "name": "频道一",
    },
  ],
  "departmentType": "provider",
  "displayName": "开单组管理员",
  "hasPermissionToAddStaff": true,
  "id": Any<String>,
  "isOnline": false,
  "profilePictureId": "1",
  "type": "manager",
  "username": "provider01",
}
`;

exports[`user get /api/v1/users/id=:userId 404 user not found 1`] = `
Object {
  "code": "888.012.404",
  "message": "用户不存在",
  "type": "NotFoundError",
}
`;

exports[`user get /api/v1/users/id=:userId 422 invalid userId 1`] = `
Object {
  "code": "888.012.422",
  "fields": Array [
    "userId",
  ],
  "message": "输入格式错误",
  "type": "RequestValidationError",
}
`;

exports[`user get /api/v1/users/id=me 200 ok 1`] = `
Object {
  "channels": Array [
    Object {
      "id": "5ec5f60b25fbb049dbd231e2",
      "name": "频道一",
    },
  ],
  "departmentType": "provider",
  "displayName": "开单组管理员",
  "hasPermissionToAddStaff": true,
  "id": "5ec4ec469d87c4387d9350ba",
  "profilePictureId": "1",
  "type": "manager",
  "username": "provider01",
}
`;

exports[`user get /api/v1/users/id=me 401 unauthorized 1`] = `
Object {
  "code": "888.010.402",
  "message": "请先登入",
  "type": "AuthenticationError",
}
`;

exports[`user patch /api/v1/users/id=:userId?via=admin 204 no content 1`] = `Object {}`;

exports[`user patch /api/v1/users/id=:userId?via=admin 403 forbidden 1`] = `
Object {
  "code": "888.012.405",
  "message": "用户无此权限",
  "type": "ForbiddenError",
}
`;

exports[`user patch /api/v1/users/id=:userId?via=admin 404 manager not found 1`] = `
Object {
  "code": "888.012.404",
  "message": "用户不存在",
  "type": "NotFoundError",
}
`;

exports[`user patch /api/v1/users/id=:userId?via=admin 422 invalid profilePictureId 1`] = `
Object {
  "code": "888.012.422",
  "fields": Array [
    "profilePictureId",
  ],
  "message": "输入格式错误",
  "type": "RequestValidationError",
}
`;

exports[`user patch /api/v1/users/id=:userId?via=manager 204 no content 1`] = `Object {}`;

exports[`user patch /api/v1/users/id=:userId?via=manager 403 forbidden 1`] = `
Object {
  "code": "888.012.405",
  "message": "用户无此权限",
  "type": "ForbiddenError",
}
`;

exports[`user patch /api/v1/users/id=:userId?via=manager 404 user not found 1`] = `
Object {
  "code": "888.012.404",
  "message": "用户不存在",
  "type": "NotFoundError",
}
`;

exports[`user patch /api/v1/users/id=:userId?via=manager 422 invalid profilePictureId 1`] = `
Object {
  "code": "888.012.422",
  "fields": Array [
    "profilePictureId",
  ],
  "message": "输入格式错误",
  "type": "RequestValidationError",
}
`;

exports[`user post /api/v1/users?via=admin 201 created 1`] = `
Object {
  "departmentType": "provider",
  "displayName": "新增主管帳號測試",
  "hasPermissionToAddStaff": true,
  "id": Any<String>,
  "profilePictureId": "1",
  "type": "manager",
  "username": "createManagerTest",
}
`;

exports[`user post /api/v1/users?via=admin 403 forbidden 1`] = `
Object {
  "code": "888.012.405",
  "message": "用户无此权限",
  "type": "ForbiddenError",
}
`;

exports[`user post /api/v1/users?via=admin 409 created 1`] = `
Object {
  "code": "888.012.409",
  "message": "帐号已存在",
  "type": "ConflictError",
}
`;

exports[`user post /api/v1/users?via=admin 422 invalid profilePictureId 1`] = `
Object {
  "code": "888.012.422",
  "fields": Array [
    "profilePictureId",
  ],
  "message": "输入格式错误",
  "type": "RequestValidationError",
}
`;

exports[`user post /api/v1/users?via=manager 201 created 1`] = `
Object {
  "departmentType": "provider",
  "displayName": "新增部門人員帳號測試",
  "id": Any<String>,
  "profilePictureId": "1",
  "shiftType": "morning",
  "type": "staff",
  "username": "createStaffTest",
}
`;

exports[`user post /api/v1/users?via=manager 403 forbidden 1`] = `
Object {
  "code": "888.012.405",
  "message": "用户无此权限",
  "type": "ForbiddenError",
}
`;

exports[`user post /api/v1/users?via=manager 409 created 1`] = `
Object {
  "code": "888.012.409",
  "message": "帐号已存在",
  "type": "ConflictError",
}
`;

exports[`user post /api/v1/users?via=manager 422 invalid profilePictureId 1`] = `
Object {
  "code": "888.012.422",
  "fields": Array [
    "profilePictureId",
  ],
  "message": "输入格式错误",
  "type": "RequestValidationError",
}
`;
